enhancements/ingress/global-admission-hsts.md

```diff
@@ -0,0 +1,726 @@
+    // If domainPatterns is specified and a route is created with a spec.host matching one of the domains, 
+    // the route must specify the annotations specified in requiredAnnotations.  If domainPatterns is empty,
+    // the specified annotations are IGNORED
+    // 
+    // The use of wildcards is allowed like this: *.foo.com matches everything under foo.com.
+    // foo.com only matches foo.com, so to cover foo.com and everything under it, you must specify *both*
+    // +optional
+    DomainPatterns []string
+    
+    // maxAge is the delta time range in seconds during which hosts are regarded as HSTS hosts.
+    // If set to 0, it negates the effect and hosts are no longer regarded as HSTS hosts.
+    // maxAge is a time-to-live value, and if this policy is not refreshed on a client, the HSTS
+    // policy will eventually expire on that client.
+    // +required
+    MaxAge                  MaxAgePolicy
+
+    // preloadPolicy directs the client to include hosts in its host preload list so that
+    // it never needs to do an initial load to get the HSTS header (note that this is not defined
+    // in RFC 6797 and is therefore client implementation-dependent).
+    // +optional
+    PreloadPolicy           PreloadPolicy
+
+    // includeSubDomainsPolicy means the HSTS Policy applies to any subdomains of the host's
+    // domain name, when the route admission policy allows wildcards.  If wildcard routes aren't allowed,
+    // this directive has no effect.
+    // +optional
+    IncludeSubdomainsPolicy IncludeSubdomainsPolicy
+}
+
+type MaxAgePolicy struct {
+    // zero means no opinion
+    // omitempty
+    LargestMaxAge int32
+    // zero means no opinion
+    // omitempty
+    SmallestMaxAge int32
+}
+
+type PreloadPolicy string
+
+var (
+    RequirePreloadPolicy   PreloadPolicy = "RequirePreload"
+    RequireNoPreloadPolicy PreloadPolicy = "RequireNotPreload"
+    NoOpinionPreloadPolicy PreloadPolicy = "NoOpinion"
+    DefaultPreloadPolicy   PreloadPolicy = ""
+)
+
+type IncludeSubdomainsPolicy string
+
+var (
+    RequireIncludeSubdomains    IncludeSubdomainsPolicy = "RequireIncludeSubdomains"
+    RequireNotIncludeSubdomains IncludeSubdomainsPolicy = "RequireNotIncludeSubdomains"
+    NoOpinionIncludeSubdomains  IncludeSubdomainsPolicy = "NoOpinion"
+    DefaultIncludeSubdomains    IncludeSubdomainsPolicy = ""
+)
+````
+A new type `RequiredHSTSPolicies` will be added to `IngressSpec` to contain any configured required HSTS Policies:
+```go
+type IngressSpec struct {
+    ...
+    // RequiredHSTSPolicies specifies HSTS policies that are required to be set on
+    // newly created routes matching some criteria.
+    // If a route specifies a HSTS annotation that matches one of the RequiredHSTSPolicies,
+    // the route is admitted.  If the route doesn't match a RequiredHSTSPolicy, the route
+    // is rejected with a message indicating the suggested policy value.
```

Comment:
Thinking about interacting with the API, I suggest a "first matching policy" rule.  Where the first matched rule is the only one that matters.  This would allow a permissive policy by default (what we currently have), that can easily be replaced with a broad, specific policy and exclusions ordered early.

What do you think?

[
  // hole punched console
  // hole punched for monitoring
  // hole punched for my special apps with ns label legacy=true
  // require all routes to have hsts set at some minimal policy
]

Reply:
It is the former, with some additional stipulations.  I updated the doc to clarify.